beacon_node/beacon_chain/src/builder.rs

```diff
@@ -278,13 +279,43 @@ where
                     .to_string()
             })?;
 
-        self.genesis_block_root = Some(chain.genesis_block_root);
-        self.head_tracker = Some(
-            HeadTracker::from_ssz_container(&chain.ssz_head_tracker)
-                .map_err(|e| format!("Failed to decode head tracker for database: {:?}", e))?,
+        let persisted_fork_choice = store
+            .get_item::<PersistedForkChoice>(&Hash256::from_slice(&FORK_CHOICE_DB_KEY))
+            .map_err(|e| format!("DB error when reading persisted fork choice: {:?}", e))?
+            .ok_or_else(|| "No persisted fork choice present in database.".to_string())?;
+
+        let fc_store = BeaconForkChoiceStore::from_persisted(
+            persisted_fork_choice.fork_choice_store,
+            store.clone(),
+        )
+        .map_err(|e| format!("Unable to load ForkChoiceStore: {:?}", e))?;
+
+        let mut fork_choice =
+            ForkChoice::from_persisted(persisted_fork_choice.fork_choice, fc_store)
+                .map_err(|e| format!("Unable to parse persisted fork choice from disk: {:?}", e))?;
+
+        let genesis_block = store
+            .get_item::<SignedBeaconBlock<TEthSpec>>(&chain.genesis_block_root)
+            .map_err(|e| format!("DB error when reading genesis block: {:?}", e))?
+            .ok_or_else(|| "Genesis block not found in store".to_string())?;
+        let genesis_state = store
+            .get_state(&genesis_block.state_root(), Some(genesis_block.slot()))
+            .map_err(|e| format!("DB error when reading genesis state: {:?}", e))?
+            .ok_or_else(|| "Genesis block not found in store".to_string())?;
+
+        let slot_clock = SystemTimeSlotClock::new(
```

Comment:
I don't think we should conjure a `SystemTimeSlotClock` from thin air here, as I think it will wreak havoc on the tests that use the testing slot clock. Instead I think we should use `self.slot_clock`, and patch up the existing couple of places where `resume_from_db` is called on the builder before `slot_clock` or `testing_slot_clock` (e.g. in `beacon_chain/src/test_utils.rs`)

Reply:
Hmm yeah agree it's going to be bad for testing slot clock. The problem is that there's a chicken-and-egg problem; we need to know the genesis time (i.e., the genesis state) start the slot clock and we need the slot clock in order to learn the canonical head.

Presently those two actions are lumped together in one function and not generic across the slot clocks... I'll see what I can do...

Reply:
I think I managed a decent solution (without changing the schema) that solves this. Let me know what you think :)

Reply:
Yep! Addresses the chicken-and-egg problem nicely!